apollo-federation/src/operation/mod.rs

```diff
@@ -1162,10 +1312,11 @@ mod field_selection {
 
     /// The non-selection-set data of `FieldSelection`, used with operation paths and graph
     /// paths.
-    #[derive(Clone, PartialEq, Eq, Hash)]
+    #[derive(Clone)]
     pub(crate) struct Field {
         data: FieldData,
         key: SelectionKey,
+        sorted_arguments: Arc<Vec<Node<executable::Argument>>>,
```

Comment:
Can we make sure `FieldData` itself always has a sorted argument list, instead of adding a copy here? If not, this is okay.

I had an idea a while back that I already experimented with a little bit. It is to have our own immutable `DirectiveList` and `ArgumentList` wrapper types that guarantee sorted order in their constructor. I can PR that in a followup.

Reply:
@goto-bus-stop 
The sorting currently is really just for faster equality/hash. It's possible we re-sort on every invocation of those, and that would take less memory since we wouldn't need to store sorted field arguments or element directives, though it'd be slower.

We could sort in-place, though that has some drawbacks:
1. The JS codebase currently leaves arguments/directives/object value fields unsorted, so we'd be diverging from that (that's more of a short-term problem though).
2. Not sorting in-place has the benefit that users basically see the order they provided in the router query, in the subgraph queries. For the case of directives, users may have particularly been relying on us not changing directive order (even though they shouldn't be), and they may view sorting as a breaking change.
3. The sort order we've arbitrarily chosen in this PR suddenly matters more, since it becomes publicly exposed, and changes in sort order influence the order in generated query plans.
4. I'm not sure what the practical memory benefit is of sorting in-place (it's certainly less memory, but I'm not sure if the amount is significant).

The 2nd point is the real issue IMO, but it'll take some time to think through/evaluate.

Reply:
Thanks, those are good points. AFAIK JS doesn't necessarily maintain all order, because of its lossy representation of object values for example. But directive order is very relevant. The `DirectiveList` and `ArgumentList` structures could account for this though and maintain two sort orders simultaneously, one for display and another one for hashing and comparison.

Anyways, that's a separate issue.

Reply:
FWIW if the **only** purpose of `sorted_arguments` is equality/hashing, we could store only a hashed `u64` in the structure instead of a whole `Arc<Vec<>>`

Reply:
With later ECMAScript versions, object properties should be in [property creation order](https://262.ecma-international.org/10.0/#sec-ordinaryownpropertykeys) (so if we've been iterating in-order, it might be fine, though not for sure there). It would be nice if `DirectiveList`/`ArgumentList` accounted for ordering concerns.

For hashing, I was curious about storing the hash in the data structure (and using `same_arguments()` for equality), though I wasn't sure how to do that properly in Rust. (It probably won't end up being significant, but definitely down to do it later if it is 😄 )